rsocket/RSocket.cpp

```diff
@@ -2,77 +2,106 @@
 
 #include "rsocket/RSocket.h"
 
-#include <folly/io/async/EventBaseManager.h>
-
 namespace rsocket {
 
-folly::Future<std::shared_ptr<RSocketClient>> RSocket::createConnectedClient(
-    std::unique_ptr<ConnectionFactory> connectionFactory,
+folly::Future<std::unique_ptr<RSocketClient>> RSocket::createConnectedClient(
+    std::shared_ptr<ConnectionFactory> connectionFactory,
     SetupParameters setupParameters,
     std::shared_ptr<RSocketResponder> responder,
     std::unique_ptr<KeepaliveTimer> keepaliveTimer,
     std::shared_ptr<RSocketStats> stats,
     std::shared_ptr<RSocketConnectionEvents> connectionEvents,
     std::shared_ptr<ResumeManager> resumeManager,
-    std::shared_ptr<ColdResumeHandler> coldResumeHandler,
-    OnRSocketResume) {
-  auto c = std::shared_ptr<RSocketClient>(new RSocketClient(
-      std::move(connectionFactory),
-      std::move(setupParameters),
-      std::move(responder),
-      std::move(keepaliveTimer),
-      std::move(stats),
-      std::move(connectionEvents),
-      std::move(resumeManager),
-      std::move(coldResumeHandler)));
+    std::shared_ptr<ColdResumeHandler> coldResumeHandler) {
+  auto createRSC = [
+      connectionFactory,
+      setupParameters = std::move(setupParameters),
+      responder = std::move(responder),
+      keepaliveTimer = std::move(keepaliveTimer),
+      stats = std::move(stats),
+      connectionEvents = std::move(connectionEvents),
+      resumeManager = std::move(resumeManager),
+      coldResumeHandler = std::move(coldResumeHandler)](
+      ConnectionFactory::ConnectedDuplexConnection connection) mutable {
+    VLOG(3) << "createConnectedClient received DuplexConnection";
+    return RSocket::createClientFromConnection(
+        std::move(connection.connection),
+        connection.eventBase,
+        std::move(setupParameters),
+        std::move(connectionFactory),
+        std::move(responder),
+        std::move(keepaliveTimer),
+        std::move(stats),
+        std::move(connectionEvents),
+        std::move(resumeManager),
+        std::move(coldResumeHandler));
+  };
 
-  return c->connect().then([c]() mutable { return c; });
+  return connectionFactory->connect().then(
+      [createRSC = std::move(createRSC)](
+          ConnectionFactory::ConnectedDuplexConnection connection) mutable {
+        // fromConnection method must be called from the eventBase and since
+        // there is no guarantee that the Future returned from the
+        // connectionFactory::connect method is executed on the event base,
+        // we have to ensure it by using folly::via
+        auto* eventBase = &connection.eventBase;
+        return via(eventBase).then(
```

Comment:
Nit: You can do `return folly::via(eventBase, [...] { ... });` and drop the `.then()`.